demos/common/python/models/ctpn.py

```diff
@@ -0,0 +1,442 @@
+        keep_inds = np.where(scores > 0.7)[0]
+        text_proposals, scores = text_proposals[keep_inds], scores[keep_inds]
+
+        sorted_indices = np.argsort(scores.ravel())[::-1]
+        text_proposals, scores = text_proposals[sorted_indices], scores[sorted_indices]
+
+        text_recs = self.text_proposal_connector.get_text_lines(text_proposals, scores, size)
+
+        heights = np.zeros((len(text_recs), 1), np.float)
+        widths = np.zeros((len(text_recs), 1), np.float)
+        scores = np.zeros((len(text_recs), 1), np.float)
+        for index, box in enumerate(text_recs):
+            heights[index] = (abs(box[5] - box[1]) + abs(box[7] - box[3])) / 2.0 + 1
+            widths[index] = (abs(box[2] - box[0]) + abs(box[6] - box[4])) / 2.0 + 1
+            scores[index] = box[8]
+        keep_inds = np.where((widths / heights > self.min_ratio) & (scores > self.boxes_threshold) &
+                             (widths > (self.min_num_proposals * self.text_proposals_width)))[0]
+
+        return text_recs[keep_inds]
+
+    @staticmethod
+    def nms(x1, y1, x2, y2, scores, thresh, include_boundaries=True, keep_top_k=None):
+        b = 1 if include_boundaries else 0
+
+        areas = (x2 - x1 + b) * (y2 - y1 + b)
+        order = scores.argsort()[::-1]
+
+        if keep_top_k:
+            order = order[:keep_top_k]
+
+        keep = []
+        while order.size > 0:
+            i = order[0]
+            keep.append(i)
+
+            xx1 = np.maximum(x1[i], x1[order[1:]])
+            yy1 = np.maximum(y1[i], y1[order[1:]])
+            xx2 = np.minimum(x2[i], x2[order[1:]])
+            yy2 = np.minimum(y2[i], y2[order[1:]])
+
+            w = np.maximum(0.0, xx2 - xx1 + b)
+            h = np.maximum(0.0, yy2 - yy1 + b)
+            intersection = w * h
+
+            union = (areas[i] + areas[order[1:]] - intersection)
+            overlap = np.divide(intersection, union, out=np.zeros_like(intersection, dtype=float), where=union != 0)
+
+            order = order[np.where(overlap <= thresh)[0] + 1]  # pylint: disable=W0143
+
+        return keep
+
+    @staticmethod
+    def generate_anchors():
+        def generate_basic_anchors(sizes, base_size=16):
+            base_anchor = np.array([0, 0, base_size - 1, base_size - 1], np.int32)
+            anchors = np.zeros((len(sizes), 4), np.int32)
+            for index, (h, w) in enumerate(sizes):
+                anchors[index] = scale_anchor(base_anchor, h, w)
+            return anchors
+
+        def scale_anchor(anchor, h, w):
+            x_ctr = (anchor[0] + anchor[2]) * 0.5
+            y_ctr = (anchor[1] + anchor[3]) * 0.5
+            scaled_anchor = anchor.copy()
+            scaled_anchor[0] = x_ctr - w / 2  # xmin
+            scaled_anchor[2] = x_ctr + w / 2  # xmax
+            scaled_anchor[1] = y_ctr - h / 2  # ymin
+            scaled_anchor[3] = y_ctr + h / 2  # ymax
+            return scaled_anchor
+
+        heights = [11, 16, 23, 33, 48, 68, 97, 139, 198, 283]
```

Comment:
I modified `generate_anchors()` according to the suggestions, also the even values were converted to odd by subtracting 1. I can delete this function either, if the hardcoded anchors are suitable in the code. Should I apply this change?
The whole postprocessing was taken from AC, so I believe the changes should be introduced there too, I will add it.

Reply:
Yes, I believe having explicit anchors is simpler thus better. Currently they are also hardcoded but implicitly through the list of `heights`. Additionally, this will eliminate the need in explanation of values of `heights`.

Reply:
Added `self.anchors` and deleted the function.